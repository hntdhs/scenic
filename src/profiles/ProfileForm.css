.edit-welcome {
    margin: 1em;
}

.edit-form {
    margin: 1em;
}

.edit-user {
    font-family: fantasy;
    color: green;
}

.profile-edit-input-label {
    font-weight: bold;
    font-size: medium;
    font-family: fantasy;
    color: green;
} 
