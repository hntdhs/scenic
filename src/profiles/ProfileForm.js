import React, { useState, useContext } from "react";
import { useHistory } from "react-router-dom";
import Alert from "../common/Alert";
import BywayApi from "../api/api";
import UserContext from "../auth/UserContext";
import LimitedTextArea from "../common/LimitedTextArea";

// eslint-disable-next-line
import useTimedMessage from "../hooks/useTimedMessage";

/** Profile editing form.
 *
 * Displays profile form and handles changes to local form state.
 * Submitting the form calls the API to save, and triggers user reloading
 * throughout the site.
 *
 * Confirmation of a successful save is normally a simple <Alert>, but
 * you can opt-in to our fancy limited-time-display message hook,
 * `useTimedMessage`, but switching the lines below.
 *
 * Routed as /profile
 * Routes -> ProfileForm -> Alert
 */

function ProfileForm() {
  const { currentUser, setCurrentUser } = useContext(UserContext);
  const history = useHistory();
  const [formData, setFormData] = useState({
    bio: currentUser.bio,
    userLocation: currentUser.userLocation,
    favoriteState: currentUser.favoriteState,
    profilePhoto: currentUser.profilePhoto,
  });
  const [formErrors, setFormErrors] = useState([]);

  // switch to use our fancy limited-time-display message hook
  const [saveConfirmed, setSaveConfirmed] = useState(false);
  // const [saveConfirmed, setSaveConfirmed] = useTimedMessage()

  /** on form submit:
   * - attempt save to backend & report any errors
   * - if successful
   *   - clear previous error messages and password
   *   - show save-confirmed message
   *   - set current user info throughout the site
   */

  async function handleSubmit(evt) {
    evt.preventDefault();

    let profileData = {
      bio: formData.bio,
      userLocation: formData.userLocation,
      favoriteState: formData.favoriteState,
      profilePhoto: formData.profilePhoto,
    };

    let username = currentUser.username;
    let updatedUser;

    try {
      updatedUser = await BywayApi.saveProfile(username, profileData);
    } catch (errors) {
      console.log('errors', errors)
      debugger;
      setFormErrors(errors);
      return;
    }

    setFormData(f => ({ ...f, password: "" }));
    setFormErrors([]);
    setSaveConfirmed(true);

    // trigger reloading of user information throughout the site
    setCurrentUser(updatedUser);
    history.push("/profile");
  }

  /** Handle form data changing */
  function handleChange(name, value) {
    // console.log(evt)
    // const { name, value } = evt.target;
    setFormData(f => ({
      ...f,
      [name]: value,
    }));
    setFormErrors([]);
  }

  return (
      <div>
        <h2>Hi there, {currentUser.username}! This is where you create and edit your profile.</h2>
        <h3>Profile</h3>
            <form>
              <div>
                <label>Bio</label>
                <LimitedTextArea value={formData.bio} limit={200} onChange={(newValue) => {return handleChange('bio', newValue)}} />
              </div>
              <div>
                <label>Location</label>
                <LimitedTextArea value={formData.userLocation} limit={100} onChange={(newValue) => {return handleChange('userLocation', newValue)}} />
                {/* <input
                    name="userLocation"
                    value={formData.location}
                    onChange={handleChange}
                /> */}
                {/* there's two different input types so might need two versions of handleChange */}
              </div>
              <div>
                <label>Favorite state to travel to</label>
                <LimitedTextArea value={formData.favoriteState} limit={100} onChange={(newValue) => {return handleChange('favoriteState', newValue)}} />
              </div>
              <div>
                <label>Profile Photo</label>
                {/* <LimitedTextArea value={formData.profilePhoto} limit={1000} onChange={(newValue) => {return handleChange('profilePhoto', newValue)}} /> */}
                <input
                    name="profilePhoto"
                    value={formData.profilePhoto}
                    onChange={(event) => {return handleChange('profilePhoto', event.target.value)}}
                />
              </div>
              {/* switched profile photo back a regular input because limited text area doesn't make sense for a URL, but that means that the handleChange is going to be a little different for the different inputs. with a regular input the onChange uses an event, not newValue */}

              {formErrors.length
                  ? <Alert type="danger" messages={formErrors} />
                  : null}

              {saveConfirmed
                  ?
                  <Alert type="success" messages={["Updated successfully."]} />
                  : null}

              <button
                  className="btn btn-primary btn-block mt-4"
                  onClick={handleSubmit}
              >
                Save Changes
              </button>
            </form>
      </div>
  );
}

export default ProfileForm;
