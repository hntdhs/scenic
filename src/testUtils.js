import React from "react";
import UserContext from "./auth/UserContext";

const demoUser = {
  username: "testuser",
  first_name: "testfirst",
  last_name: "testlast",
  email: "test@test.net",
  photo_url: null,
};
// const
let UserProvider =
    ({ children, currentUser = demoUser }) => (
    <UserContext.Provider value={{ currentUser }}>
      {children}
    </UserContext.Provider>
    )


// const UserProvider =
//     ({ children, currentUser = demoUser, hasAppliedToJob = () => false }) => (
//     <UserContext.Provider value={{ currentUser, hasAppliedToJob }}>
//       {children}
//     </UserContext.Provider>
// );

export { UserProvider };
