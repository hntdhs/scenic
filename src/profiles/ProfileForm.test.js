import React from "react";
import { render } from "@testing-library/react";
import ProfileForm from "./ProfileForm";
import { UserProvider } from "../testUtils";

let browser;
let page;

beforeEach(async () => {
  browser = await puppeteer.launch({headless: true, executablePath: '/Applications/Google Chrome.app/Contents/MacOS/Google Chrome'});
  page = await browser.newPage();
});

afterEach(async () => {
  await browser.close()
});

it("updates the profile page on submission", async () => {
  
    await page.goto('http://localhost:3000/profile');
  
    await page.type('#bio', 'integration testing');
// problem - when I add an id to the submit button in the JSX, it makes the snapshot test fail
  
    await Promise.all([
      page.click('#save'),
      await page.waitForNavigation()
    ]);
    await expect(page.url()).toMatch('http://localhost:3000/username');
    await expect(page.url()).toContain('integration testing');
// question - how do I determine username?
// what would be the way to test if what I added for bio is appearing on the profile page - expect(page.url).toContain('')
  
  });




// it("login succesfully", async () => {
  
//     await page.goto('http://localhost:3000/login');
  
//     await page.type('#password', 'atlantic');
//     await page.type('#username', 'atlantic');
  
//     await Promise.all([
//       page.click('#loginSubmit'),
//       await page.waitForNavigation()
//     ]);
//     await expect(page.url()).toMatch('http://localhost:3000')
  
//   });