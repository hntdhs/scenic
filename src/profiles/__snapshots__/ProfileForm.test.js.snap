// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <h2>
      Hi there, testuser! This is where you create and edit your profile.
    </h2>
    <h3>
      Profile
    </h3>
    <form>
      <div>
        <label>
          Bio
        </label>
        <div>
          <textarea />
          <p>
            Maximum length: 200
            <br />
            0/200
          </p>
        </div>
      </div>
      <div>
        <label>
          Location
        </label>
        <div>
          <textarea />
          <p>
            Maximum length: 100
            <br />
            0/100
          </p>
        </div>
      </div>
      <div>
        <label>
          Favorite state to travel to
        </label>
        <div>
          <textarea />
          <p>
            Maximum length: 100
            <br />
            0/100
          </p>
        </div>
      </div>
      <div>
        <label>
          Profile Photo
        </label>
        <input
          name="profilePhoto"
          value=""
        />
      </div>
      <button
        class="btn btn-primary btn-block mt-4"
      >
        Save Changes
      </button>
    </form>
  </div>
</DocumentFragment>
`;
